src/Microsoft.DotNet.Build.Tasks/GatherFoldersToRestore.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Text;
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class GatherDirectoriesToRestore : Task
+    {
+        [Required]
+        public string[] RootDirectories { get; set; }
+
+        [Output]
+        public string[] DirectoriesToRestore { get; set; }
+
+        public int DirectoryListLength = 7000;
+
+        public override bool Execute()
+        {
+            HashSet<string> directoriesToRestore = new HashSet<string>();
+
+            foreach (string rootDirectory in RootDirectories)
+            {
+                AddDirectoriesToRestore(rootDirectory, rootDirectory, directoriesToRestore);
+            }
+
+            List<string> resultBuilder = new List<string>();
+            StringBuilder sb = new StringBuilder(DirectoryListLength);
+
+            foreach (string directoryPath in directoriesToRestore)
+            {
+                sb.Append('"');
+                sb.Append(directoryPath);
+                sb.Append("\" ");
+
+                if (sb.Length >= DirectoryListLength)
+                {
+                    resultBuilder.Add(sb.ToString());
+                    sb.Clear();
+                }
+            }
+
+            if (sb.Length > 0)
+            {
+                resultBuilder.Add(sb.ToString());
+            }
+
+            DirectoriesToRestore = resultBuilder.ToArray();
+
+            return true;
+        }
+
+        public void AddDirectoriesToRestore(string directoryToExplore, string directoryToAdd, HashSet<string> directoriesToRestore)
+        {
+            bool hasProjectJson = File.Exists(Path.Combine(directoryToExplore, "project.json"));
+
+            if (hasProjectJson)
+            {
+                directoriesToRestore.Add(directoryToAdd);
+            }
+
+            foreach (string childDirectory in Directory.GetDirectories(directoryToExplore))
+            {
+                AddDirectoriesToRestore(childDirectory, hasProjectJson ? childDirectory : directoryToAdd, directoriesToRestore);
```

Comment:
If we expect to keep this long term, I'm happy to move to that model. I initially added it to work around the exact DNU issues were were seeing, with the understanding that we would remove this once we were off DNU and on to NuGet.

There's a bit of a cost in moving away from this model, since it means that we basically just take all the project.json's and then split them in to groups (this task becomes conceptually simpler since it can remove all the file system logic and just take in an item group of paths and a limit and return an item group of strings of paths where they are combined up to that limit).

If it didn't make the common case slower (see my comment a little later about having to invoke dotnet restore five times instead of once), I'd be all for it.  As is, I'm a little hesitant, but I'll do whatever you and @dagood would prefer.